### ISSUE TITLE
error: mixing declarations and code is incompatible with standards before C99 [-Werror,-Wdeclaration-after-statement]

### ISSUE THREAD

**Issue by mumin16** [android/ndk#1718]
### Description

/home/x64/android-ndk-r25-beta4/sources/android/cpufeatures/cpu-features.c:942:9: error: mixing declarations and code is incompatible with standards before C99 [-Werror,-Wdeclaration-after-statement]
    int regs[4];

### Affected versions

r25

### Host OS

Linux

### Host OS version

arch linux

### Affected ABIs

x86_64

### Build system

CMake

### minSdkVersion

26

**Comment by mumin16** [comment]
next line give same error(int vendorIsIntel). I think we should put the variable declaration at the beginning of the function.

fixes the problem that is:

static void
android_cpuInit(void)
{
    int regs[4];
    int vendorIsIntel;

**Comment by jfgoog** [comment]
This is almost certainly not a bug. ANSI C requires that all variables be declared at the beginning of a function, and you are probably passing command-line flags to clang to enforce it.

For example:

```C
#include <stdio.h>
int main() {
  printf("Hello, world\n");
  int foo = 0;
  return foo;
}
```

With NDK 24.0.8215888:

* `toolchains/llvm/prebuilt/linux-x86_64/bin/clang foo.c` works fine.
* `toolchains/llvm/prebuilt/linux-x86_64/bin/clang -ansi -pedantic -Werror foo.c` produces the error you described.
* `toolchains/llvm/prebuilt/linux-x86_64/bin/clang -std=c99 -pedantic -Werror foo.c` also works fine, because since C99, declarations and code may be mixed.

**Comment by DanAlbert** [comment]
That is our code that has the issue, fwiw :)

But the answer to anything in the NDK's cpufeatures is that it's old and bad and what you actually want is https://github.com/google/cpu_features. The only reason we continue to ship ours is to avoid breaking existing users. The fix here is trivial, but I'm not sure if polishing a turd actually helps? I'm guessing this failure actually helped you discover https://github.com/google/cpu_features (I'm not advocating introducing new failures, to be clear)

**Comment by enh-google** [comment]
is it normal that the deprecated cpufeatures would be built with the _user's_ cflags? is that fixable? (otherwise it seems like there are lots of ways to break cpufeatures' build via cflags :-) )

/me wonders whether we should have a `#warning "deprecated; use https://github.com/google/cpu_features instead"` in cpufeatures?

**Comment by DanAlbert** [comment]
> is it normal that the deprecated cpufeatures would be built with the user's cflags?

ndk-build users can just import the module and get our flags (though if they add these flags to `APP_CFLAGS`, it'd still infect the module). Everyone else _must_ provide their own flags because they have to build it somehow.

> /me wonders whether we should have a #warning "deprecated; use https://github.com/google/cpu_features instead" in cpufeatures?

Might be a good idea, but it'd be difficult to ack the warning and move on if you don't care (and if the user is fine with it, I don't really care, I just want them to know that there are better options). I suppose we could guard the warning with something like `#ifdef NDK_CPUFEATURES_DEPRECATION_ACK` to make it easy to ignore for `-Werror` builds?

**Comment by enh-google** [comment]
yeah, i hadn't really thought of that. especially if they're not necessarily already controlling the cflags for that part of the build (otherwise they could just "not -Werror that code").

**Comment by DanAlbert** [comment]
I'm also not sure if a deprecation warning would even be visible to Android Studio users. I don't think the warning will be visible unless the file also has an error.

**Comment by enh-google** [comment]
i do notice that there's nothing about cpu-features (as opposed to cpufeatures) in the cpufeatures directory: not in the .mk file, the .h file, or the .c file. and there isn't even a README.md file at all yet. should we add something to one/all of these?

**Comment by DanAlbert** [comment]
I thought we had done that already, but you're right. Maybe we talked about it but forgot to follow through? https://android-review.googlesource.com/c/platform/ndk/+/2112126
